File: core/conn/security_dll/native/source/Key.cpp
/**********************************************************************
// @@@ START COPYRIGHT @@@
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// @@@ END COPYRIGHT @@@
//
**********************************************************************/

#include "Key.h"
#include "StaticLocking.h"

Key::Key()
{
   m_key = NULL;
   //m_kType = PUBLIC;
   m_keyLen = 0;
}

Key::~Key() 
{
	if (m_key)
		EVP_PKEY_free(m_key);
}

void Key::getPubKeyFromFile(const char *cert_file) throw (SecurityException)
{
   X509* cert = openCertFile(cert_file);
   
   if (!cert)
       throw SecurityException(ERR_OPEN_INPUT_FILE, (char*)cert_file);

   //Get expiration date of the certificate
   strncpy((char*) m_certExpDate, (char*) (X509_get_notAfter(cert)->data), EXPDATESIZE);
   m_certExpDate[EXPDATESIZE] = '\0';

   m_key = X509_extract_key(cert);

   if (!m_key)
   {
     X509_free(cert);
	 throw SecurityException(ERR_RETRIEVE_KEY_FROM_FILE, (char* )cert_file);
   }

//   m_kType = PUBLIC;
   m_keyLen = EVP_PKEY_size(m_key);
   X509_free(cert);
}

X509* Key::openCertFile(const char* certFile) throw (SecurityException)
{
   if (!certFile)
   {
	   throw SecurityException(INPUT_PARAMETER_IS_NULL, " - certFile.");
	   return NULL;
   }

   FILE* inFile=fopen(certFile, "r");
   if (!inFile)
   {
	   throw SecurityException(FILE_NOTFOUND, (char* )certFile);
	   return NULL;
   }

   getMutex();
   X509 *certificate = PEM_read_X509(inFile, NULL, 0, NULL);
   releaseMutex();
   if (!certificate)
   {
	   fclose(inFile);
	   throw SecurityException(ERR_READ_CERT_FILE, (char* )certFile);
	   return NULL;
   }  

   fclose(inFile);

   return certificate;
}

EVP_PKEY* Key::getPrivKeyFromFile(const char* inFile,
								  EVP_PKEY *priv_key) throw (SecurityException)
{
   if (!inFile)
	 throw SecurityException(INPUT_PARAMETER_IS_NULL, " - inFile.");

   FILE* file=fopen(inFile, "r");
   if (!file)
      throw SecurityException(FILE_NOTFOUND, (char *)inFile);

   EVP_PKEY* key = PEM_read_PrivateKey(file, &priv_key, 0, NULL);

   fclose (file);
   if (!key)
      throw SecurityException(ERR_RETRIEVE_KEY_FROM_FILE, (char *)inFile);

   return key;
}

unsigned char* Key::getCertExpDate(const char* certFile) throw (SecurityException)
{
   X509* cert = openCertFile(certFile);
   
   if (!cert)
	   throw SecurityException(ERR_OPEN_INPUT_FILE, " - cert.");

   unsigned char* date = X509_get_notAfter(cert)->data;

   X509_free(cert);

   return date;
}


EVP_PKEY* Key::getKey()
{
   return m_key;
}

void Key::setKey(EVP_PKEY* key)
{
	m_key = key;
}

int Key::getKeyLen() 
{
   return m_keyLen;
}

unsigned char* Key::getCertExpDate() 
{
   return m_certExpDate;
}

File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/Key.java
/**********************************************************************
// @@@ START COPYRIGHT @@@
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// @@@ END COPYRIGHT @@@
//
**********************************************************************/

/**
 * Key.java
 */

package org.trafodion.jdbc.t4;

import java.io.FileInputStream;
import java.io.InputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.security.cert.X509Certificate;
import java.security.KeyFactory;
import java.security.interfaces.RSAPublicKey;
import java.security.interfaces.RSAPrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import javax.crypto.KeyGenerator;


public class Key {
	public Key(){}

	/** Reads the public key from the certificate file and
	 *  stores the key and the length of the public key
	 *  in member data.
	 *  @param X509Certificate - The certificate stored
	 *  the public key
	 */
   public void getPubKeyFromFile(X509Certificate cert)
   {
      m_pubKey = (RSAPublicKey) cert.getPublicKey();
      if (((m_pubKey.getModulus().bitLength()) / 8) > 128)
    	  m_pubKeyLen = 256;
      else
    	  m_pubKeyLen = 128;
   }

   /** Reads the private key from certificate file and
    *  stores the key in the member data.
    *  @param String - The file stored the private key
    *  @throw SecurityException
    */
   public void getPrivKeyFromFile(String inFile) throws SecurityException
   {
      InputStream inStream=null;

      try{
         // Loading private key file
         inStream=new FileInputStream(inFile);
         byte[] keyBytes=new byte[inStream.available()];
         inStream.read(keyBytes);
         inStream.close();

         // Read the private key from file
         PKCS8EncodedKeySpec privKeySpec=new PKCS8EncodedKeySpec(keyBytes);
         KeyFactory keyFactory = KeyFactory.getInstance("RSA");
         m_privKey= (RSAPrivateKey) keyFactory.generatePrivate
                                                       (privKeySpec);

      }catch (FileNotFoundException fnf) {
         throw new SecurityException(fnf, SecClientMsgKeys.FILE_NOTFOUND, new Object[]{inFile});
      }catch (IOException io) {
         throw new SecurityException(io, SecClientMsgKeys.ERR_OPEN_INPUT_FILE, new Object[]{inFile});
      }catch (Exception e) {
         throw new SecurityException(e, SecClientMsgKeys.ERR_RETRIEVE_KEY_FROM_FILE, new Object[]{inFile});
      }finally {
         try {
            if (inStream != null)
               inStream.close();
         }catch (IOException io) {
            // not much we can do at this point
         }
      }
   }

   /**
    * Generates a secret key using AES algorithm and 128 bits key
    * @param sessionKey the session key byte array used for symmetric key
    *                   generation
    * @return the SecretKey
    * @throws SecurityException
    */
   static SecretKey generateSymmetricKey(byte [] sKey) throws SecurityException
   {
	  if (sKey == null)
		  throw new SecurityException(SecClientMsgKeys.INPUT_PARAMETER_IS_NULL, new Object[]{"sKey"});
	  try {
         // Get the KeyGenerator
         KeyGenerator kgen = KeyGenerator.getInstance("AES");
         synchronized(kgen) {
        	 kgen.init(128);
         }
         // Use the lower 16 bytes of the session key to generate the 128 bits secret
         // key used for data encryption

         SecretKey skey = new SecretKeySpec(sKey, kgen.getAlgorithm());

         return skey;
	  }catch (NoSuchAlgorithmException nae) {
		  throw new SecurityException(nae, SecClientMsgKeys.ERR_CREATE_SYMMETRIC_KEY, null);
	  }
   }

   /**
    *
    * @return the public key
    */
   public RSAPublicKey getPubKey()
   {
      return m_pubKey;
   }

   /**
    *
    * @return the private key
    */
   public RSAPrivateKey getPrivKey()
   {
      return m_privKey;
   }

   /**
    *
    * @return the length of the public key
    */
   public int getPubKeyLen()
   {
      return m_pubKeyLen;
   }

   private RSAPublicKey m_pubKey;
   private RSAPrivateKey m_privKey;
   private int m_pubKeyLen;
}
